### ISSUE TITLE
Expression / Formula editor doesn't respect units.

### ISSUE THREAD

**Issue by Connor9220** [FreeCAD/FreeCAD#16861]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

If you go into a parameter box and click the formula / expression icon in the upper right hand corner and type in a number without a unit (such as in or " for example), it returns back with the proper number in the main parameter box, however, when you okay out, it converts the number to metric.  If you simply type in the number in the normal parameter box without going into the expression / formula editor, it respects the default units as expected.  This looks to affect multiple workbenches including sketcher and CAM.

### Full version info

```shell
OS: Debian GNU/Linux 12 (bookworm) (GNOME/gnome)
Word size of FreeCAD: 64-bit
Version: 1.0.0RC2.38806 (Git) AppImage
Build type: Release
Branch: (HEAD detached at 1.0rc2)
Hash: 3d63fc6c2f665a8d5e6468845a419bcac80756c7
Python 3.11.9, Qt 5.15.13, Coin 4.0.3, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: OpenDark.qss/OpenDark/Qt default
Installed mods: 
  * OpenTheme 2024.9.1
  * OpenDark 2024.3.13
  * lattice2 1.0.0
  * ThreadProfile 1.92.0
  * fasteners 0.5.28
```

### Subproject(s) affected?

None

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by benj5378** [comment]
For me, putting a Spreadsheet cell `=(5 mm)^2)` is given as `6,25 cm^2` on my wayland device, and on my non-wayland it's given as `25 mm^2`.

We might also mention:  
* #13534 
* and #12173
* idk if #11862 is Expression or Spreadsheet issue

**Comment by Connor9220** [comment]
Still an issue.

**Comment by Machinist-May** [comment]
If you use a scientific calculator this isn't an issue. It's only when using the software to calculate your math that this happens. It's not ideal.

**Comment by pieterhijma** [comment]
I'm investigating https://github.com/FreeCAD/FreeCAD/issues/30826 and came across the following behavior in FreeCAD 1.1 (the last version before #28371) that was unknown to me:

If you use schema `US Customary` and you create a document with a VarSet with a property with type Length and you don't add any value, then the value of this property is `0 in`.

If you then change `0 in` to `23` (note that there is no unit), you get as value: `1.92'` which is 23 in / 12 feet. So, what happens is that 23 is interpreted as inches and converted to feet.

Then if you change `1.92'` to `34` (again no unit) you get `11.33 yd` which is 34 feet / 3 yards.

Then if you change `11.33 yd` to `45` you get `45 yd`.

In other words, the value editor remembers its last unit and interprets a value without unit in terms of the remembered value.

The same is the case in the metric system: If you switch to that system and you change `0 mm` to `10000` you will get `10 m` as a result. If you change `10 m` to `30` you will get `30 m` as a result.

**Comment by maxwxyz** [comment]
Reopened due to https://github.com/FreeCAD/FreeCAD/pull/31392

**Comment by tritao** [comment]
I looked into this briefly as part of the hackaton, since I'm already looking into a sourrounding area in https://github.com/FreeCAD/FreeCAD/pull/30139.

This is an analysis that explains the issue and how we can fix it cleanly as part of a follow-up PR:
https://gist.github.com/tritao/6610fe63917cd1651d2e1e95c5254849

**Comment by pieterhijma** [comment]
Great, thanks, @tritao!
